### PR TITLE
pd: fix local_rank and in mutlti nodes training

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
1. get local rank from `PADDLE_LOCAL_RANK` environment variable.
2. disable synchronization in forward-backward and synchronize manually before opt.step()

### CODE REVIEW

File: doc/train/parallel-training.md [hunk 1]
```diff
@@ -218,6 +218,21 @@ NUM_WORKERS=0 HDF5_USE_FILE_LOCKING=0 python -m paddle.distributed.launch \
...
+# ----- train_pp.sh -------
+unset CUDA_DEVICE_MAX_CONNECTIONS
+python -m paddle.distributed.launch --gpus="0,1,2,3,4,5,6,7" --log_dir logs dp --pd train input_torch.json -l train_pp.log
+# -------------------------
+```
+
+Then, run the script on the first node with:
+
+```bash
+mpirun run_pp.sh
```
Comment (njzjz) [inline doc/train/parallel-training.md:233]: Should the number of ranks (`-n 8`) be assigned?

File: deepmd/pd/utils/env.py [hunk 2]
```diff
@@ -27,7 +27,8 @@
     ncpus = os.cpu_count()
 NUM_WORKERS = int(os.environ.get("NUM_WORKERS", min(0, ncpus)))
 # Make sure DDP uses correct device if applicable
-LOCAL_RANK = paddle.distributed.get_rank()
+LOCAL_RANK = os.environ.get("PADDLE_LOCAL_RANK")
+LOCAL_RANK = int(0 if LOCAL_RANK is None else LOCAL_RANK)
```
[the remark this answers is not in the document]
Comment (caic99) [reply deepmd/pd/utils/env.py]: @HydrogenSulfate Would you help me applying this patch, thanks.

File: deepmd/pd/train/training.py [hunk 3]
```diff
@@ -741,16 +743,30 @@ def step(_step_id, task_key="Default") -> None:
...
-                        **input_dict,
-                        cur_lr=paddle.full([], pref_lr, DEFAULT_PRECISION),
-                        label=label_dict,
-                        task_key=task_key,
-                    )
 
-                with nvprof_context(enable_profiling, "Backward pass"):
-                    loss.backward()
+                # disable synchronization in forward-backward manually
+                # as derivatives exist in model forward
```
Comment (caic99) [inline deepmd/pd/train/training.py:748]: @HydrogenSulfate What's the point here? Sorry but I still can't get it.

**Review by caic99 — APPROVED** [review]

File: deepmd/pd/train/training.py (hunk 3 above)
Comment (HydrogenSulfate) [reply deepmd/pd/train/training.py:748]: > @HydrogenSulfate What's the point here? Sorry but I still can't get it.

The DDP in Paddle might introduce some gradient synchronization hooks during the backward computation, which are unnecessary for higher-order differentiation. Therefore, we manually disable these synchronization hooks here and then manually synchronize the gradients.

PyTorch does not seem to have this issue.
image

File: doc/train/parallel-training.md (hunk 1 above)
Comment (HydrogenSulfate) [reply doc/train/parallel-training.md:233]: > Should the number of ranks (`-n 8`) be assigned?

This parameter will be automatically inferred by `paddle.distributed.launch` and the subsequent parameters.
